### DISCUSSION TITLE
Probing correctly with iosender

### DISCUSSION THREAD

**Discussion by Klippert** [terjeio/grblHAL#197] [Q&A] [answered]
First of all, thanks for a brilliant piece of software.

I'm not sure weather this should be in IO sender or Grbl-Hal, but here goes.

I'm struggling a bit to understand how the probing and tool offset works. 
I have a workbee cnc with limitswitches and was used tro openbuilds control and GRBL.
So i set up grbl hal on one of Phils boards, and I'm now able to home the machine and set G54 to my home position.
I've set the probe plate position, and offset so the machine runs off to G53.9 and does a probing and says it has the tool length offset (i can see it in the console, but that's about the only place?)

Then i just want to do a simple facing of a piece of POM, and i set X and Y of G54 to the corner of that piece. 

Now to set Z on G54, i would normally use the probed position from my touch plate, which i know is precisely 25mm above my base plate. And the Gcode facing i've made with work coordinate system starting on the baseplate. 

Now i'm just missing the connection between the last two dots... - how do i get from the probed tool length to the surface of my spoilboard?

Is it a seeting in Grbl-hal or is it in IO sender? and where do i need to read to fill in my gaps?

Best regards
Kristoffer

**Comment by S2000Stefan** [comment]
Hello Kristoffer,
I didn't quite understand what you wanted to know, but I'll give you a short workflow for a tool change that might help you.

First you need to set a reference point at your preferred position on the machine (G59.3) by doing the following in the probing tab:
In the tool length offset, check Establish reference offset and start a probing.
![Referenz1](https://user-images.githubusercontent.com/35189051/104130275-0f1e0400-5370-11eb-86c2-7c3b83d62cff.jpg)

After successful probing, you will receive the following message:
![Fertig2](https://user-images.githubusercontent.com/35189051/104130295-34127700-5370-11eb-8c88-2de1f57dfb40.jpg)

And in the GRBL tab, a diode lights up at TLO ref'd.
![Anzeige3](https://user-images.githubusercontent.com/35189051/104130341-6d4ae700-5370-11eb-83ec-1c13eeb34b78.jpg)

After that, as always, you do a probing of your touchplate on your spoilboard, I like to use a macro for that.
It could look like this:
G38.2 Z-10 F20

After the successful probing you have to enter your height manually in the DRO. In your case 25,000mm.  
![Eingabe25_4](https://user-images.githubusercontent.com/35189051/104130593-da12b100-5371-11eb-9327-50667ce07d7b.jpg)

Then, for each new tool you want to use, you can move to your reference position (G59.3) and probing with a $TPW system command and ioSender will convert the height.
However, this requires that you set a tool change mode in the settings that is suitable for you. 
![tool-chance-mode-5](https://user-images.githubusercontent.com/35189051/104130791-0e3aa180-5373-11eb-92fa-aaae55e0cd43.jpg)
Unfortunately, this is only a very brief description, but I hope that it has helped you a little.

**Reply by Klippert** [reply]
Thanks - that explains the flow very clearly! 

a couple of questions arose:
1. You do the initial probing with any old tool in i expect?
2. the baseplate probing - does that have to be done as a macro and a probing? or could i just identify the tip of the tool touching the baseplate and add 25 to that coordinate? (
3. Ideally i'd want to use automatic touch off - that's why I have the fixed touch probe set up - so once i have the height of z-home over my baseplate, any tool should just be touched off, and good to go. - in a perfect world.

- so this is how far i got: 
1. touched off on the touchplate with a tool, however the selection was "none"
2. i went down to the baseplate and set that point to 0 in the dro (just where the tool touches the plate)
3. loaded a file and start cycle and it runs correctly. 
4. changed tool, and did a probe at G59.3 with tool 1, and it just keeps using the same offset as before. 

At the moment i have tool specific files, so this only ran the process with the current tool, which worked like a charm.
If i create a file with toolchanges in, i guess i need to select "manual touch off"?
And i guess it will pause and allow me to manually change the tool, before running off and touching it off and starting the path?

Best 
Kristoffer

**Comment by S2000Stefan** [comment]
> Ideally i'd want to use automatic touch off - that's why I have the fixed touch probe set up - so once i have the height of z-home over my baseplate, any tool should just be touched off, and good to go. - in a perfect world.

O.k. if you want to use a fixed probe, the procedure is as follows.
Anyway, that's what I do:

- All three linear axes (X,Y,Z) should be homed.
- G59.3 must be set up for the tool length sensor, is also the tool change position.
- After loading the milling program, the number of tool changes contained in the program is displayed. 
It is also advisable, if the first tool change is known, to select the first tool in the MDI window with M61Qxx or in the Tool Tab, where xx stands for the tool number. This is not necessary, it only speeds up the subsequent sequence.
- First, the reference point must be set by approaching the tool length sensor, touchplate with the first tool.
- In your case, you now enter 25 in the dro because you know that your fixed length sensor is 25mm away from the spoilboard.
Of course, the cutter must touch the fixed length sensor at that moment. e.g. with G38.2 Z-2 F10
- When everything is set up, the milling job can be started.
- Since we have already set up the first tool, the first tool change command in the job is skipped.
If a tool change in the milling programme is initiated by an M6 Txx (where xx stands for the tool number), ioSender automatically switches the milling spindle off and moves to Z max in order to then move to the set G59.3 position to change the tool. 
- After the tool change, move briefly over the tool sensor with jogging commands downwards in order to carry out probing with the system command $TPW in order to determine the tool length offset which the ioSender automatically calculates.
- Press the start button to continue running the nc programme until the next tool change, repeat the logical points above. 
  
In the above case, I would set the settings to "Manual touch off @ G59.3". 
In my opinion, macros make work much easier, and in the ioSender macros are very easy to set up, but of course they are not a must.
Of course, you also have to set up the postprocessor for the tool change.


+---------------------------------------------------
+  Commands output at toolchange
+---------------------------------------------------

begin TOOLCHANGE

"M5" (milling spindle is switched off)
"M9" (switch off mist and flood, optional)  
"T[T]M06(MSG,[TOOLNAME])" (tool change)
"M3[S]" (spindle start) 
"M07" (switch on mist, optional)
"M08" (switch on flood, optional)  
"G4 P5" (pause to bring the spindle up to speed)


Tomorrow I can also explain how to set up your fixed tool length sensor in the probing tab so that you can initiate a automatic probing.
:-)

I forgot to tell you, I would also recommend that you lower the speeds until the process is working properly so that you can intervene if there are any problems.

**Reply by Klippert** [reply]
Hello again,

Ok, i got time to test it out, and I sort of got something working. not entirely. 

Just to clarify, I'm using Beta 8 of Io sender. 

I don't quite it. 
Tell me where i'm going wrong, please. 
My touch probe is at G53.9. 
It is exactly 26mm above my spoilplate/baseplate. (i know i wrote 25mm earlier but that was a typo)
so any tool that touches off, should be exactly 26 mm above the baseplate. 
Therefore I don't get why I need to tell IO sender anything but the fact that it's that high?
And then every time a tool is changed, I can just touch off and go. 
Technically Z shouldn't even care if it was homed? (as long as the tooltip is above the touchplate.


Here is what happens:
- Load the gcode, and io sender tells me there are 2 toolchanges, and Tool Lenght reference should be set.
- I tell it that it has tool 1 in by M6Q1 (or the tool dropdown under work parameters)
- If i use the Probing and check "Probe fixture @G59.3 with tool 1 - it goes to the probe plate, probes the tool, and then runs back to where it was, without raising the tool off the probe first. - which is thought wasn't quite great as i nearly broke a tool on traveling across the baseplate at full speed in a height of 26mm. - and then at the end it raises the Z axis. 
- Now, i have a z value in the DRO that is not really telling me much. however i have a button appearing under tool length offset saying "clear tool length offset".
- The TLO ref'd is not lit up.
- If i do the same with both "Probe fixture ... " and "establish reference offset" checked if goes to the probe, does a probing, again slides the probe off the plate and runs back to where it came from. 
- Now tool TLO ref'd is lit. But i dont' have any clue what Z value i'm at.
- So i have to jog to the plate, do a manual probe with G38.2Z-2F20 (i learned something new ;-)) 
- Now i get a Z value, that i know is 26.000mm above the baseplate. - so i go to the DRO and enter 26 in the Z field, and it bounces right back to where it was... 
- I click probing, and Io-sender crashes with no error messages. 
- I start Io-sender again, and it still says my tool is TLO ref'd. and DRO has the same values, so i guess it was all in controller still. lucky me ;-)
- I load the gcode file again, and enters 26 in the Z field in the DRO - it now works. (Bug?)
- So now i know where my baseplate is, and i can run the first part of the job. As long as i remember to raise the z-axis first :-) 
- When the toolchange comes up, it runs up to the front, and ask me for a tool change, which i do, and when i then use "probe fixture @59.3..." it get the right height and i can run the rest of the program.

Now my big question is - why can't I just use "fixture heights (which i've set to 26mm) under probing, and probe G59.3 once and have the software update the DRO automatically? 

Ok, so this got rather longwinded, and appologies for my lack of understanding. If anybody read this far, you deserve a few pics of the machine as well :-)
https://www.grbl.org/workbee-based-router

**Comment by S2000Stefan** [comment]
O.K. Let's try again.  :-)

- First, G59.3 must be set up for the tool length sensor, is also the **tool change position**. 
You can do this by selecting G59.3 in the GRBL window under Work parameters. Then move to your fixed tool length sensor with jog commands and zero x and y. Raise the z-axis to a height that allows you to change the tool. Zero the Z-axis **at this height** as well. (G59.3)

- After loading the milling program, the number of tool changes contained in the program is displayed.
It is also advisable, if the first tool change is known, to select the first tool in the MDI with **M61Qxx** or in the Tool Tab, where xx stands for the tool number. This is not necessary, it only speeds up the subsequent sequence.
As you have already realised, you can also use the tool dropdown under work parameters to select the tool.

- Then select e.g. G54 in the GRBL window under Work parameters.
- Now clamp the appropriate cutter and jog briefly approx 5mm over your fixed tool length sensor at position G59.3.
- In the Probing tab, tick Establish reference offset, nothing else, if it is not already there and press start at the bottom. Now a probing is carried out and the reference point is set. You should also see in the GRBL tab that TLO ref'd is lit.
- After that, you have to touch your fixed touch plate **AGAIN** with G38.2Z-2F20. I'll tell you later how to do it more comfortably when it works for you. And enter 26mm in the DRO at Z. (G54) 
- Then jog to your workpiece and set your X and Y zero point (G54). This is also easier, but more about that later.
- If everything is now set up correctly, you can start the nc programme.
- Since we have already set up the first tool, the first tool change command in the job is skipped.
If a tool change in the milling programme is initiated by an M6 Txx (where xx stands for the tool number), ioSender automatically switches the milling spindle off and moves to Z max in order to then move to the set G59.3 position to change the tool.
- After the tool change, move briefly over the tool sensor with jogging commands downwards in order to carry out probing with the system command **$TPW in the MDI in order to determine the tool length offset** which the ioSender automatically calculates. The second TLO led then lights up.
- Press the start button to continue running the nc programme until the next tool change, repeat the logical points above.
- Don't forget to set "Manual touch off @ G59.3" in the settings.
- And as I said, I would lower the speeds to be able to intervene if necessary if something is not working properly.
- Hope it works better this time.  ;-)

**Reply by Klippert** [reply]
Hello again, thanks for your patience ;-)

Here's what happens:
- I home the machine
- I check that GRBL-Hal is set to "manual touch off @ G53.9" - it is :)
- Jog to G53.9 and set x,y and z to 0 when it's above the touchplate. (that was actually already done)
- Open the .nc, which has two toolchanges, and io-sender tells me that there are two toolchanges, and tool length refernce should be established before start.
- interestingly, under work parameters I can not select tool numbers (in my case tool 15 and tool 18)
![image](https://user-images.githubusercontent.com/53910032/104437843-94cbca80-558f-11eb-8430-efb029512516.png)
- In the MDI i enter M61Q15 - which is the first tool.
- now work parameters has tool 15. 
- I switch to G54 on work paramters
- I go to probing and select only Establish reference offset. 
![image](https://user-images.githubusercontent.com/53910032/104438192-ff7d0600-558f-11eb-8fc7-611e9637ff45.png)
- and get a probing completed.
- And the spindle goes up. nothing more. 
- TLO ref'd is now lit under work parameters.
- I enter 26 in Z in the DRO under grbl tab - and it did not crash :)
![image](https://user-images.githubusercontent.com/53910032/104438782-ac578300-5590-11eb-9515-17081c27bedd.png)
- start cycle, and it runs. It does look like i've been a bit too deep, as the bit goes a bit into my baseplate.
- At the end of the toolpath it comes up to a block Pending T18 M6 having moved to over the probe plate all nicely. 
- I switch tool, and do a $TPW and it does a probingand goes back up a bit from the probe. 
- this time i decided to see where i was with my Z 0 in G54. so i jogged to a free space on the build plate, and gently lowered the tool, and do the old fashioned test with a piece of paper. and it would seem that my Z is approximately 0.9mm too low. so i must have messed something up, when entering the z-26 in the dro? or something like that.
- I jegged z-up and started the last part of the program and it finished nicely.

So, the big question is, why does the first probing give a wrong offset?
I could try doing the probing manually, set Z=0 and then moving over to the baseplate and see if the height of the baseplate is actually -26 still. (it should be.)

And if I reset the controller and home it again, will it forget the referece points?

Oh, and one more question, if i press the emergency stop it stops everything, but when i restart, is there a way to restart only from a specific tool? It seems to restart from the beginning of the whole nc program, which in some cases is quite far from where i want to be.


Best regards and thanks for your patience - I seem to be learning  ;-)

**Comment by S2000Stefan** [comment]
I am very pleased that the tool change is now almost working properly. We will also get rid of the height offset. :-)
> interestingly, under work parameters I can not select tool numbers (in my case tool 15 and tool 18)

Only four tools can be selected, but by entering M61Qxx, any other tool is possible.

> So i must have messed something up, when entering the z-26 in the dro? or something like that.

Yes, I now suspect that you have not touched your fixed tool length sensor twice, as I wrote, or the height is not exactly 26mm.
I hope I understood you correctly that you want to have Z-zero on the top of your Machine bed?
If so, I have also seen that you have already entered your 26mm in Touch plate/fixture heights.
Let's try this after these points:

- and get a probing completed.
- And the spindle goes up. nothing more.
- TLO ref'd is now lit under work parameters.

You don't have to make an entry in the DRO, the Touch plate/fixture heights can also automatically scan ioSender in the following way:

- The reference point is set correctly and you are still standing short above your fixed touchplate, standing on the machine bed. Now go to the Probing tab under Edge finder, external, click in the middle of the graphic, then click start. Now ioSender will re-probe and automatically transfer the 26mm you entered in the Touch plate/fixture heights window to your DRO Z-axis.

![Edge-finder](https://user-images.githubusercontent.com/35189051/104513052-59f97f00-55ef-11eb-99d9-fd4b1be0d2a4.jpg)

In the Edge finder, external you can also very comfortably touch the corners or edges of your workpiece, but make your attempts very carefully and at very slow speeds until you have understood how it works. In the left area you can set the speeds, traverse paths and tool diameters. 

> And if I reset the controller and home it again, will it forget the referece points?

I'll take that as a given.  It's probably also a safety issue.  ;-)

> Oh, and one more question, if i press the emergency stop it stops everything, but when i restart, is there a way to restart only from a specific tool? It seems to restart from the beginning of the whole nc program, which in some cases is quite far from where i want to be.

In the case of an emergency stop, I assume that something has gone wrong and the workpiece is no longer usable anyway. in any case, I only use the emergency stop when it is really necessary. It's probably also a safety issue.  ;-)

**Reply by Klippert** [reply]
Thanks again for your very nice description. Lets see if i get it all right.
The probing was a bit of a aha-moment for me :-)

Ok - here we go - episode 4...

First i did a test to see if my touchprobe is actually 26mm, as it should be. 
1. I home the machine
2. God to touchprobe and do a probing using the edge finder and probe the and set the coordinate system G54 with no offset:  (after adjusting my probe height to 25mm as i was 1 mm off)
![image](https://user-images.githubusercontent.com/53910032/104846772-84438900-58dc-11eb-8b86-cf4ab0a9a219.png)
Then jog to a place on the spoilboard, and amazingly enough Z=0 when the bit is just touching the baseplate. So the offset error must have been my own brainfart remembering the touchprobe height wrong. So that works as it should now.

And I seem to have understood the edgefinder external when probing and using the touchplate height. 

So, now it's probing correctly. And when I do the Tool length Offset (establish refernece offset) i get the tool changes working too. 

Not for one or another weird reason, when i go to the tool length offset tab and just want to probe the fixture @53.9 with tool 29 (as that's the first one), it moves exactly 1mm to the left (x-1) and just sits there. It used to be that it'd go to the probe and do a probing.

All i get in the console is this:
$G
[GC:G0 G54 G17 G21 G91 G94 G49 G98 G50 M5 M9 T29 F0 S0.]
ok
Probing...
G91F200
ok
PM:ok
G53G1F600X1Y519.155
ok
PM:ok
G53G1F600Z3
PM:cancel
G91
Probing cancelled/failed
ok
ok
$G
Probing failed
[GC:G0 G54 G17 G21 G91 G94 G49 G98 G50 M5 M9 T29 F0 S0.]
ok
G91
ok

(the cancel i triggered by pressing cancel as it was just sitting there doing noting. )

I tried restarting the teensy and iosender but with no effect.
I wonder what have changed, since it's no longer behaving.

But the good thing is, i now seem to understand the probing ;-)

/Kristoffer

**Comment by terjeio** [comment]
_Probe fixture @ G59.3_ in the _Tool length offset_ tab is broken, looking at that now.

> I wonder what have changed, since it's no longer behaving.

Is the G59.3 offset set correctly? Does it work correctly on a tool change (assuming $341 mode is set for G59.3 touch off)?

**Reply by Klippert** [reply]
yep. it was set up correctly. and it funnily enough went to the probe on a toolchange in the program. Anyhow it seems that we found a bug. :)

**Comment by terjeio** [comment]
New [edge version](http://io-engineering.com/downloads/) uploaded.

I have changed movements for `Probe fixture @ G59.3` in the `Tool length offset` tab to better correspond to controller G59.3 mode tool change movements.

When no tool reference offset is established (or to be established) `Probe fixture @ G59.3` will set tool offset so that the _Fixture_ height  (in the left panel) becomes the probed position (including all other offsets).  Please comment if this is a bad idea.

Other changes are that probe XY offsets can be set as part of profiles, these fields are hidden unless the height of the window is slighly increased. A new option for keeping the last window size has been added to the "Settings: App tab to avoid doing that on every startup.

**Reply by Klippert** [reply]
The link above to the new edge version seems to be broken. 
I'm a bit unclear if this means a new version of grbl-hal or a new version of io-sender?

Either way i'm very keen to try it out

**Reply by terjeio** [reply]
Link is fixed now. It is a new version of the sender only.

**Reply by Klippert** [reply]
ah. it was needing  www. in front. http://www.io-engineering.com/downloads/

**Comment by S2000Stefan** [comment, marked as answer]
> When no tool reference offset is established (or to be established) Probe fixture @ G59.3 will set tool offset so that the Fixture height (in the left panel) becomes the probed position (including all other offsets). Please comment if this is a bad idea.

Since I don't use the automatic probing with Probe fixture @ G59.3, since G59.3 is also my tool change position and the probing path is quite long with shorter tools, Kristoffer should comment on how he finds it, since he wants to use it more often.
I tested it very briefly today and it seems to work. :-)

The other new features, on the other hand, I find very useful for me.

![Mpos1](https://user-images.githubusercontent.com/35189051/104946581-972c8b00-59ba-11eb-9591-4b5dc33ee47b.jpg)
:+1: 
![Backup](https://user-images.githubusercontent.com/35189051/104946677-cb07b080-59ba-11eb-8b5e-5a12f41fcb2e.jpg)
:+1: 
![Fenster](https://user-images.githubusercontent.com/35189051/104946801-fbe7e580-59ba-11eb-86f4-253b19a11139.jpg)
:+1: 

I compiled the firmware with Optimize most (-O3) as suggested in another thread, which saves some flash memory, (-O0) 82.6kb (-O3) 49kb, whereby I even received a warning message with (-O0) that no longer occurred with (-O3).  ;-)
For your information @terjeio.

**Reply by Klippert** [reply]
Ok - something is working very well. i have run into a few adjustment problems with my HBS-56stepper drivers, that i need to solve, before i can properly test the probing, but it does seem that the probing is working as expected now. and it actually does go to G53.9 and does the probing as expected.

**Reply by Klippert** [reply]
Ok, now i'm getting some weird behaviour that I haven't had before. 
I click home, and Z first homes (touches sensor twice) then y and x homes also touching the sensors twice, and home becomes lit up in green. 
when i click probe (probe fixture...) i get an error saying i have to home first.
![image](https://user-images.githubusercontent.com/53910032/105402353-9c424200-5c27-11eb-8a1e-09a8237b0d33.png)

I tried restarting IO-sender and re homing and voila - it worked. I think I might have tried probing without having homed the machine first, and then the error state was not reset by homing, but i'm not sure about  that.

Now i can set the Touch plate/Fiture height, and whenver i click probe fixture, the input box nicely changes to "fixture" under fixture height. very nice that i'm able to have both a fixure height and a touchplate height.

So here goes what i've done and what happened :-)
First i've homed the machine,
Set the fixture height to 25mm - and chose probe fixture and establish reference offset. 
So now G54 Z0 should be the surface of my baseplate. - hoever it isn't - my baseplate is at -19mm. 
So i go do a manual probing - jog to over the fixture, and use the edge finder, external. 
![image](https://user-images.githubusercontent.com/53910032/105403639-466e9980-5c29-11eb-9e6e-b12fb5d42392.png)
That does the job, the G54 Z0 is now at the surface

I did think this probing would have set G54 as well.
![image](https://user-images.githubusercontent.com/53910032/105403899-a402e600-5c29-11eb-97ed-484f7abb890c.png)

Never the less, the tool is now TLO ref'd and it seems i'm ready to rock :)


I've loaded up a program with 6 toolchanges. (a bigger job)
It starts by going to machine Z0  (raising the tool) - goes to the fixture and waits for the first tool change (well the starting tool that is). (as it should.). I probe with $TPW, and it probes the tool, and I click cycle start 

After the toolpath complted and the toolchange comes up, it goes up, stops the spindle, comes the whole way up, and moves over the fixture, and awaits my toolchange. Toolchange done, I do a $TWP to probe the length, and it does the probing and wait for cycle start.

And the program nicely continues its job. 

Unfortunately my pc figured things had gotten boring after running for about half an hour and decided to go to sleep. so the spinle just stopped in place and waking things up did nothing. so a strong recommendation to switch off all "sleep" features if you're running on windows.

I'll give things another shot tomorrow.

**Reply by terjeio** [reply]
Which version of grblHAL are you running?

_Help > About > To Clipboard_ and paste here (or to a file for upload).

> Set the fixture height to 25mm - and chose probe fixture and establish reference offset.
So now G54 Z0 should be the surface of my baseplate. - hoever it isn't - my baseplate is at -19mm.

When both _Probe fixture @ G59.3_ and _Establish reference offset_ is checked `Fixture height` is greyed out as it is not used.
A `G49` is sent to clear any active tool offset and a `$TLR` command is issued to set the reference.

If only _Probe fixture @ G59.3_ is checked _Fixture height_ is enabled and a `G43.1` is sent to set the current tool length offset so that the probed position becomes the _Fixture height_ - this includes all currently active offsets (`G5x` and `G92`).

> I did think this probing would have set G54 as well.

It does not when `Add offset` is unchecked.

**Reply by Klippert** [reply]
![image](https://user-images.githubusercontent.com/53910032/105533271-2e138300-5cec-11eb-9ba7-e62c5c91b69d.png)

Ok, so that all seems to work as it should with probing. Or maybe more correctly, now I seem to understand how it\s working. :-)
One questions that remains:
When i had an "accident" in the middle of a program (ie. my pc fell asleep, or i triggered the emergency stop, for one reason or another), but my machine and workpiece is actually ok, and the positions are intact, what is the easiest way of resuming operations? - In this case it would be nice to be able to just home the machine, go back a little and continue from there, but i can't seem to find any other way than editing the G-code, and even if i had a line number, that would be hard, as i don't know where i could continue from. Is there a clever way around that problem?

**Reply by phil-barrett** [reply]
I think we all have had those situations where you stop in the middle of a job for what ever reason and want to restart.  I edit the GCode and drop the blocks between the preamble and the most recent rapid prior to the problem.  Assuming you were running in absolute mode and did not lose your zeros, it should be fine. You might need to restart spindle and reset feed rate, some CAMs are skimpy with S and F codes.

I suppose there is a feature request in there but I think it is a hard problem to solve generally (what is a preamble? Where is there a safe starting point?).  Would be fantastic, if possible...

**Reply by Klippert** [reply]
Yes - It is a tricky problem to solve, as one never really knows where would be appropriate to start from. I guess what i was hoping for was some sort of way of dividing the gcode into segments that would then not be "forgotten" but kind of stored in a log or something, so at least one would know where one got to.

**Reply by Klippert** [reply]
And by the way, a really warm thanks to all of you who has helped me this far! You guys are amazing :)

**Reply by Klippert** [reply]
I'm still getting the error about homing before probing, when i first home, and then try to probe. 
![image](https://user-images.githubusercontent.com/53910032/105572555-1d005b80-5d58-11eb-80c8-354fb24364ec.png)

If i close Io Sender and open it again, load the file again i don't get the error. This strikes me as a bug somewhere?

**Reply by terjeio** [reply]
> If i close Io Sender and open it again, load the file again i don't get the error. This strikes me as a bug somewhere?

Yes, there is a bug, I have just uploaded a new edge version that should fix it.

**Reply by Klippert** [reply]
I downloaded the latest version of iosender, and tested the probing. I can see that the probing of fixture works and establishing reference offset work. however i can't seem to find a way to get it to set G54 without having to jog to the probe fixture first. It would be great if i was able to use "probe fixture @G59.3" and do the workpiece offset and thereby set G54. 

Am i missing something?

**Reply by terjeio** [reply]
> Am i missing something?

Maybe I am, there is a bewildering array of use-cases.

> however i can't seem to find a way to get it to set G54 without having to jog to the probe fixture first.

You can enter the required value in the DRO? I'll have to see how I can enable _Add offset_ when both _Probe fixture @G59.3_ and _Establish reference offset_ is checked.

**Reply by Klippert** [reply]
Sorry for making things complicated. 
My problem is: When I home the machine and do a probing at the fixture at G53.9 - i can not get to the height of my baseplate. Weather i use Probe fixture with or without Establish reference offset. I am left without a Z0 being the height of my base plate, despite the fixture height is entered in the fixture height. 

Does this make sense?

The difference for me, is that i have the fixture in the opposite end of the machine compared to my home position, so jogging there takes a lot longer than if i can just click probe fixture.

I mean we are not hunting for the million dollar feature here ;)

Thanks again, 
Bestregards
Kristoffer

**Reply by terjeio** [reply]
> Does this make sense?

I am a bit slow. You want Z0 to become the height of the baseplate _and_ set the reference tool offset so that subsequent tool changes touched off @ G59.3 still has Z0 as the height of the baseplate?

If so then I have to allow _Probe fixture @G59.3_, _Establish reference offset_ and _Add offset_  beeing checked at the same time as well as the _Fixture_ field beeing left enabled when so. But then _Workpiece height_ has to be disabled or it will get even more confusing? Or better when both _Probe fixture @G59.3_ and _Establish reference offset_ is checked hide the _Workpiece_ offset box and only enable and show the _Action_ box inside that.

**Reply by Klippert** [reply]
Well. let me try to be a little more specific. I think i left some holes in my explaination.

I would like G54 Z0 to be set as the height of my baseplate, as i almost always use my baseplate as Z reference point in my drawings. That makes it very easy for me to go from fusion 360 and get the CNC started. 

Perhaps it would be easier if the "set coordinate system G54" was outside the workpiece offset box and just under establish reference offset? - 
That way you'd chekc, "I want to probe the fixture @G59.3", "i want to establish reference offset" and "I want G54 Z0 to be 0 based on the fixture height" 
I should add, that i'm no user-experience expert. ;-)

This would then allow me to Probe, get the TLO and G54Z0 set and be ready to cut in one operation. 

Thanks very much for your patience.

**Reply by terjeio** [reply]
The latest [edge version](http://io-engineering.com/downloads) has changes to _Tool length offset_ probing that I hope is ok, please be careful when using this and report any issues!

**Reply by Klippert** [reply]
Brilliant! 

Now here is what happens:
I home the machine.
I click probe fixturee and establish reference offset
the machine does a probing, and the G54 Z0 is now set to the baseplate height. (it wasn't before)

Now for a further test, i manually set the Z in the DRO to be 10mm above the baseplate.
I click probe and set coordinate system or offset. 
![image](https://user-images.githubusercontent.com/53910032/106352546-60bd0d00-62e4-11eb-9394-35f290246094.png)

and hit start. 
It goes to the fixture, does a probing, and again my  G54 and DRO Z0 is now the height og my baseplate.

Either way gives me the behaviour I was after, though i'm not sure what the difference is when i click "Set coordinate system or offset" ?

I find this a fantastic improvement ;-)
Thank you very much!

/Kristoffer

**Reply by terjeio** [reply]
> i'm not sure what the difference is when i click "Set coordinate system or offset"

When checked it sets the selected coordinate system or G92 instead of setting the tool length offset. If _Establish reference offset_ is checked then the current tool offset is cleared and not part of the new offset, _if not it will be_. The main difference is that the tool offset is not persistent and is cleared on a soft reset , G5x and G92 are as these are saved across sessions too (G92 persistence is dependent on[ compatibility level](https://github.com/terjeio/grblHAL/wiki/Compatibility-level), if >1 it is not saved). So a tad complicated.

At some point this tab will be used to bind tool offsets to tools in a persistent tool table in the controller, with this is place it will be possible to switch (indexed) tools without touch off. For automatic tool changers (ATC), the grblHAL core is ready for adding plugins for these...

**Reply by Klippert** [reply]
That makes sense! Thanks for the explanation.

I really like the idea of indexed tool lengths stored in GRBL-HAL, as I do have a fancy dream of adding some sort of atc at one point or another. I've spotted that Mechatron has a "add on" atc, that you can add to "any" 80mm spindle. only it is Iso10, so max collet size is ER16, where i'd very much like to get to ER20 so i can use my 12mm tools.
